equinox-shipping/Domain/FinalizationProcessManager.fs

```diff
@@ -0,0 +1,56 @@
+module FinalizationProcessManager
+
+    open FSharp.UMX
+    open Types
+
+    type Service(transactions : FinalizationTransaction.Service, containers : Container.Service, shipments : Shipment.Service) =
+
+        // Transaction ID can be a md5 hash of all shipments ID
+        let execute (transactionId : string<transactionId>) : FinalizationTransaction.Events.Event option -> Async<bool> =
+            let rec loop (update: FinalizationTransaction.Events.Event option) =
+                async {
+                    let loop event = loop (Some event)
+
+                    let! action =
+                        transactions.Apply(transactionId, update)
+
+                    let untag (ids: string<shipmentId>[]) : string[] = ids |> Array.map UMX.untag
+
+                    match action with
+                    | FinalizationTransaction.Action.AssignShipments (containerId, shipmentIds) ->
+                        // TODO: This needs throttling on Cosmos.
+                        let! result =
+                            seq {
+                                for sId in shipmentIds -> async {
+                                    let! res = shipments.Execute(sId, Shipment.Command.Assign containerId)
+                                    return (sId, res)
+                                }
+                            }
+                            |> Async.Parallel
+
+                        let (successes, failures) =
+                            Array.partition snd result
+
+                        return!
+                            match Array.map fst failures with
+                            | [||] -> loop (FinalizationTransaction.Events.AssignmentCompleted {| shipmentIds = untag shipmentIds |})
```

Comment:
there's no reason they can't be of type [alias] `ShipmentId` - that'd be the point and benefit